### ISSUE TITLE
Decouple User Security and Contacts

### ISSUE THREAD

**Issue by andythorne** [sulu/sulu#3377]
| Q | A
| --- | ---
| Bug? | no
| New Feature? | yes
| Sulu Version | ^1.5.0
| Browser Version | n/a

#### New Feature

I think the user setup is a bit odd at the moment. We have a `user`, essentially just a credential set, which is tied 1-1 with a `contact` via a field in the `user` entity. I don't think the the SecurityBundle and components should be user-source agnostic and not know anything about contacts.
For example, at the moment authenticating with a different user provider (for example LDAP, facebook or a user api) requires extensively modifying the security bundle to handle user creation, modification and retrieval.
As a note, content nodes have an assigned contact (as an author). If we keep and enforce a 1-1 relationship between users and contacts, we can ensure content nodes have a blameable audit record.

##### Users
- Decouple users and contacts completely. The Security bundle should not know or care about a contact.
- Decouple user and role/permission structure from SecutiryBundle. The SecurityBundle should work with interfaces so any authentication (user) and authorisation (role/permission) implementations can be used.
- Create new interface and permission set for user administration. 

##### Permissions
- Secured controllers must always carry an entity context and must carry an entity id context when working with specific entities (get, put, post, patch). Entity + id context will allow document owners to access their content (e.g. users can access their AND their role's objects via the API).

#### Summary

This will require multiple changes across several bundles so the likelihood of a breaking change is quite high, but I still feel it's work doing.

Suggestions and feedback welcome 😄

**Comment by danrot** [comment]
I am still missing the concrete use case you are working on: Do you want to handle the login and the entire permission system via LDAP or something similar? I am asking because with your suggestion the `SecurityBundle` would almost only be a set of interfaces, and we would have to implement our permission logic in another bundle, introducing another layer, which again adds a lot of complexity.

And the other thing that worries me even more is that in our domain we assume that every user has a contact, and we want to have stuff like the first and last name also on the user. So I can see a few options here:

1. We completely split the SecurityBundle and ContactBundle. That means that a contact as an author has always to be attached to a page, because we don't have the necessary information on the user.
2. We have an optional dependency on the contact referencing a user. That means that we don't force a user to have a contact, and we would have to use something like the user name as fallback.
3. We duplicate the information we would like to have on the user (so there would be a first/last name on the user and contact). This also adds additional complexity, since we probably would have to add a check in templates do distinguish between users and contacts. Maybe we could introduce an interface for the common stuff, but then we would still have to handle these fields twice (once in the contact and once in the user).
4. Split the SecurityBundle and ContactBundle, and introduce a third bundle, which acts as a bridge between these two. The third bundle could then introduce a required field on either the Contact or the User, which then means that we have a direct link guaranteed again. A problem might be that introducing this bundle in an existing project might break the consistency of the database (because the links do not exist yet, but there have to be a link to be consistent). This would have to be documented very clearly.

Although of the drawback I would favor option 4. It also allows you to ditch the SecurityBundle completely, and replace it with an LDAP implementation. This kind of modularity might get even more interesting in case we start adapting Symfony Flex.

**Comment by andythorne** [comment]
I think the real use case is to be able to have sulu authenticate and authorize with ANY system, both of which could be completely separate. For example, you can have LDAP authentication and local db permissions based on the LDAP user reference; or (in our case) a central User API used across many (non-sulu) projects that allows you to set permissions for the project in question (via contexts) in the User API itself (kind of think of it like the current user/permissions via roles sulu structure, but over an API).

As such, I think you should be able to substitute the user provider (like you would normally in a symfony project via the secrutiry.yml user provider) and sulu should be able to play well with that. At the moment, if you change the user provider, sulu will break.

I'm not saying sulu should provide an implementation for LDAP/AD/<a system>, just make such a thing possible via interfaces (already done in part, but there are references to the Security and Contacts bundles in them).

By doing this and removing all SecurityBundle components from the ContactsBundle, we can keep the current "contact as an author" structure. The contacts bundle can then reference a user (via either a reference/username or id - whatever the user provider determines is a unique identifier is for a user) 

Given this, I think either 1 or 4 are both perfectly valid options. By favouring option 1, we keep the SecurityBundle; with 4, we will be making it replaceable.

**Comment by danrot** [comment]
Thanks for your input. I think @chirimoya would agree that this is something worth considering in general, without deciding on the exact solution.

However, this won't be possible to be done before Release 2.0. I also think this is a lot of work and I am not sure if we are even able to do that in 2.0, since we already have a lot of other focuses for that release, which would mean it has to wait until our 3.0 release, because that's probably a quite heavy BC break...

If you are willing to put a lot of time into this, you can try to make this work, and ask for our assistance at any time. If the end result is satisfying for us, we might merge it into the 2.0 release. But I think this a big, challenging task that requires a lot of coordination with us.

**Comment by andythorne** [comment]
Yep I 100% agree that this should be a major version release. The BCs would be significant. 

I haven't got a problem working on this. As usual, time will be the limiting factor here as it is, as you say, a big job. If we can get an agreed approach and a spec together, I'd be happy to do it.

**Comment by danrot** [comment]
Okay, the thing is that we are about to start working on 2.0, and there will be a bigger change in the frontend part, so it wouldn't make sense to start working on the frontend right now. We also have to discuss with our designer how that would look in terms of UX.

I hope it is ok if we can come back to you later, after we have built an opinion about that?
